dev-docs/RFCs/v7.3/program-manager-rfc.md

```diff
@@ -0,0 +1,125 @@
+# RFC: Program Management in luma.gl
+
+* **Author**: Tarek Sherif
+* **Date**: July, 2019
+* **Status**: **Draft**
+
+
+## Summary
+
+This RFC specifies subsystem for managing and re-using GLSL programs.
+
+
+## Background
+
+Creation and management of programs is currently handled by several systems in luma.gl:
+- Modules are simply JavaScript objects that can be created anywhere by the application.
+- Injections and shader hooks are managed as library-wide global state managed by functions `setShaderHook` and `setModuleInjection`.
+- Assembling base shader sources, modules, defines and injections into a final shader source is handled by `assembleShaders`.
+- Caching and re-use of programs is handled by `ShaderCache`, which is currently limited in that it scopes caches to `BaseModel` objects meaning that `BaseModel`s cannot share programs (this is due to an easily-remedied limitation in the `BaseModel` class that it doesn't set uniforms per-draw).
+
+Issues with the current system include:
+- These systems are currently only consolidated around the `BaseModel` class, which treats programs as immutable. This means that modifying programs and taking advantage of these system involves re-constructing the entire `BaseModel`. This complicates interaction with programs as they and parameters that modify their source must generally be treated as immutable and separated out from mutable properties (e.g. the [data filter layer extension](https://github.com/uber/deck.gl/blob/6113d2c8984c406e9df59c16f19630a18f36c42d/modules/extensions/src/data-filter/data-filter.js#L42) in deck.gl has to handle `filterSize`  and `softMargin` separately from other props because they modify the shader code).
+- Shader hooks and module injections are global state meaning that an application that imports luma.gl must use the same definitions globally (a case where one might not want this is when rendering to separate canvases on the same page).
```

Comment:
> a case where one might not want this is when rendering to separate canvases on the same page

I assuming this concerns using multiple gl contexts. Since support for multiple gl contexts is used in this RFC and the comments as a motivation for the proposed design perhaps be more clear about what issues we would need to solve to support multiple context canvases:

- The main issue is that we need to cache a separate set of Program instances for each gl context? 
- I.e. not that shader hooks or other parts of the shader definitions need to be defined differently per context?

Reply:
It just seems like an odd constraint to me that all systems on a page that use luma.gl would have to share these definitions.

E.g. say I was rendering a HUD with one AnimationLoop, and the scene with another, each with their own gl, do they really need to share the same hooks?